### PR TITLE
adjacency lists are now sorted

### PR DESCRIPTION
**Pull request by sbromberger**

### CODE REVIEW

**Comment by sbromberger** [comment]
@CarloLucibello @jpfairbanks this is a fairly significant merge. Could you please review? Thanks.

**Comment by CarloLucibello** [comment]
Sorry, I didn't show up  lately. I've been travelling. I'm not a huge fan of this PR, since keeping a vector dinamically sorted sounds not natural to me, vectors are just not made for this. Have you done more benchmarking?

I've gone throught the code and i could spot no fallacies

Somewhere you said that with this we could also have edge indexing. How are the two related? edge indexing would be a nice addition to LG

Side note: lately I've been taking a close look to the implementation of an adjacency list graph here https://git.skewed.de/count0/graph-tool/blob/master/src/graph/graph_adjacency.hh. If you are familiar with C++ it is worth taking a look for some inspiration. graph-tool is a nice library, fast and packed with features. See here for an overlook (also here you can see what I meant by graph vies/filters) https://graph-tool.skewed.de/static/doc/quickstart.html

At line 336 of graph_adjacency.hh you can see the internal members of the graph class, it is very simple and in next few lines you can get an idea of how edge indexing is implemented

**Comment by sbromberger** [comment]
The dynamically-maintained adjacency lists actually improve operations (at the expense of insertion; see #231 comments for the hairy details). The nice thing about sorted vectors is that we can close #20 efficiently, and we can also do your edge iteration more efficiently as well (since we can immediately know when s > d and go to the next source).

Edge indexing will also benefit from sorted vectors for the same reason edge iterations are: you know when you can move to the next adjacency vector.

**Comment by CarloLucibello** [comment]
>The dynamically-maintained adjacency lists actually improve operations (at the expense of insertion; see #231 comments for the hairy details, but insertions are now O(log deg) instead of O(1))

O(deg) I think, not  O(log deg). Edge removal instead remains O(deg)

>PS: I'm a fan of graph-tool and have communicated with its author a couple of times. My goal with LG is a pure-Julia, one-language solution, of course. :)

Of course, I was wondering how the same implementation would perform in pure julia

**Comment by sbromberger** [comment]
> O(deg) I think, not O(log deg). Edge removal instead remains O(deg)

I thought that `searchsorted()` was logN. Then we have the splice which is more cost, but I don't know how much.

**Comment by CarloLucibello** [comment]
searchsorted is log N. I don't know what exactly does splice, but if you have to insert an element in the middle of a vector you have to resize the vector and move O(d) elements of one position

**Comment by sbromberger** [comment]
@CarloLucibello if we don't keep vertices sorted within adjlists, how do we make efficient `add_edge()` and edge index functions? (Looking for ideas here!)

**Comment by sbromberger** [comment]
Multiedges are a whole different level of pain that have major downstream effects. I don't think we're ever going to support them in LG.

As far as "4 ints per edge" being memory efficient, that's the opposite direction of where I think we're going. With this solution we've got one int per (directed) edge (assuming zero overhead of the adjacency list itself).

**Comment by CarloLucibello** [comment]
edge sets don't give you indexing, and they improve performance of has_edge only in dense graphs, so that was wasted space in my opinion. On the other hand I would sacrifice some space for edge indexing functionality.

The getindex implementation I see in getindex is O(N), therefore not very usefull as it is, and also it seems wrong to me in graphs, it indenxes directed ednges insteaf of edges. 

The same function could be implemented with O(log N) complexity at the extra memory cost of allocating something similar to a cumulated degree distribution

**Comment by sbromberger** [comment]
We can get edge indexing to O(deg), I think, and possibly O(log deg). The getindex implementation is wrong - it double-counts edges - and it's likely not performant. See https://groups.google.com/d/msg/julia-users/mc8BmPW2c9A/WgKYyHcKBgAJ for more discussion.

**Comment by sbromberger** [comment]
So, I think it's O(deg) if we do something like this:

```
function getindex(g::Graph,i::Int)
    s = 1
    di = 1
    while i > 0
        # println("i = $i, s = $s, di = $di, g.fadjlist[s][di] = $(g.fadjlist[s][di])")
        if s > g.fadjlist[s][di] || di >= length(g.fadjlist[s])
            s += 1
            di = 1
        else
            di += 1
        end            
        i -= 1
    end
    if di == 1
        s -= 1
        di = length(g.fadjlist[s])
    else
        di -= 1
    end
    return Edge(s,g.fadjlist[s][di])
end
```

File: src/digraph.jl [hunk 1]
```diff
@@ -76,22 +76,23 @@ end
...
+    _insert_and_dedup!(g.fadjlist[s], d)
+    _insert_and_dedup!(g.badjlist[d], s)
     push!(g.edges, e)
     return e
 end
 
 
 function rem_edge!(g::DiGraph, e::Edge)
-    reve = reverse(e)
     has_edge(g,e) || error("Edge $e is not in graph")
```
Comment (CarloLucibello) [inline src/digraph.jl]: this is ok for now but once we get rid of the edge this  can be eliminated  in favour of ```i != 0 || error("Edge $e is not in graph")``` a few lines later

Comment (sbromberger) [inline src/digraph.jl]: yup, that's really nice. Can you make sure to check that once we get rid of edge sets? :)

File: src/core.jl [hunk 2]
```diff
@@ -1,5 +1,8 @@
 abstract AbstractPathState
 
+# taken from http://stackoverflow.com/questions/25678112/insert-item-into-a-sorted-list-with-julia-with-and-without-duplicates
+_insert_and_dedup!(v::Vector{Int}, x::Int) =
+    (splice!(v, searchsorted(v,x), [x]); v)
```
Comment (CarloLucibello) [inline src/core.jl]: it is not really important, but ```(splice!(v, searchsorted(v,x), x); v)``` is also ok

Comment (sbromberger) [inline src/core.jl]: This is actually an old version. I'm currently using

```
_insert_and_dedup!(v::Vector{Int}, x::Int) = isempty(splice!(v, searchsorted(v,x), x))
```

**Pull request merged**